### ISSUE TITLE
[Question] BlackBox Matrix-Matrix Inference

### ISSUE THREAD

**Issue by preritt** [pytorch/botorch#753]
# 🚀 Feature Request
BlackBox Matrix-Matrix Inference (BBMM) for Gaussian Process

Are there plans to implement BBMM in BOTorch to speed up matrix inversion and scalability?

Thanks

**Comment by Balandat** [comment]
Are you referring to the methods described in the GPyTorch paper (https://proceedings.neurips.cc/paper/2018/hash/27e8e17134dd7083b050476733207ea1-Abstract.html)? These have been implemented in gpytorch from the start and are therefore available in BoTorch.\

**Comment by preritt** [comment]
Yes, this is what I am referring to. Sorry, I did not know about it.

Is there a tutorial that illustrates how to use it?  For instance, I am using the code below. How can I ensure the fast matrix inversion is implemented using BBMM?


```
  model = SingleTaskGP(train_X=train_inputs_raw_GA_torch, train_Y=train_outputs_raw_GA_torch,likelihood=likelihood
                       )
  mll = ExactMarginalLogLikelihood(model.likelihood, model)
  fit_gpytorch_model(mll);
```
Thanks!

**Comment by Balandat** [comment]
For fitting the model you can pass `approx_mll=True` to `fit_gpytorch_model` so approximate log prob computations are used (note that even in gpytorch this only is used for large training data sets since for smaller problems the cholesky based actually turns out to be faster). 

The other place where this matters is for performing posterior solves. We don't set anything specifically in terms of the gpytorch options, so it should default to the LinearCG solves for large enough training data. For smaller models gpytorch will again use Cholesky by default.

**Comment by preritt** [comment]
Thanks for getting back. I tried with `approx_mll=True`and it works without any error.

However, in GPyTorch, there are multiple ways for approximate inference. Is there a way to select which one we are interested in, or is BBMM used be default?

[Approximate GP Inference](https://docs.gpytorch.ai/en/v1.1.1/marginal_log_likelihoods.html#approximate-gp-inference)

Thanks,

**Comment by Balandat** [comment]
Well Approximate GP inference here is a whole different model altogether and uses variational GP models. You can use these with the BoTorch APIs as well. We have some code for this, we could make a PR to put this up in BoTorch. 

What is the reason you want to use approximate inference? Do you have a ton of data? In general it can be challenging to get variational GPs to "work well" in Bayesian optimization due to their calibration often not being great (underestimating the posterior variance) so unless you need to use it it's better to start with Exact / other scalable GP models.

**Comment by preritt** [comment]
Thanks for your response. 

Right now I am working on smaller problems to test BoTorch. However, there are some bigger problems with more than 100 inputs, that I'd like to use this on. The number of data instance keeps increasing over time as I am using `Active learning` thus increasing the complexity of matrix inversion. I was wondering if approximate inference could be helpful to speed up the computation when dataset size increases over time. 

Does BoTorch use the link in [variational GP](https://docs.gpytorch.ai/en/v1.1.1/examples/04_Variational_and_Approximate_GPs/SVGP_Regression_CUDA.html) when setting the flag to true? 

Thanks

**Comment by Balandat** [comment]
> there are some bigger problems with more than 100 inputs, that I'd like to use this on

So note that since the GP covariances are of size `n x n` for `n` train points regardless of the input dimension `d` an approximate GP will not help there per se. But if, as you suggested, you get a lot more observations (large `n`) then it could. 

> I was wondering if approximate inference could be helpful to speed up the computation when dataset size increases over time.

Yes. The case of conditioning the model on additional observations sequentially is something than can also be treated in more sophisticated ways than just throwing a scalable GP at the full data without "looking back". The recent paper  https://arxiv.org/abs/2103.01454 outlines such an approach. Luckily, that paper implemented the method in GPyTorch and already tested it with BoTorch (https://github.com/wjmaddox/online_gp). Not sure if @wjmaddox has any plans to integrate this with the GPyTorch/BoTorch libraries...

> Does BoTorch use the link in variational GP when setting the flag to true?

You mean the `approx_mll=True` flag? No, that just controls whether GPyTorch will use Lanczos based log likelihood estimates. To use the variational GP in BoTorch you'll have to define a new model. This is pretty straightforward but requires some understanding of the `Model` API. As I mentioned above, this is something we could put out as a canned model if this is useful. The main advantage of a variational GP is that you can use minibatching for training. If that's not necessary it also make sense to look at other scalable approaches for exact GPs (e.g. those in https://github.com/cornellius-gp/gpytorch/tree/master/examples/02_Scalable_Exact_GPs)

**Comment by preritt** [comment]
Thanks a lot for the helpful references - these are helpful.
If the model for variational GP can be implemented, that would be helpful too.

Thanks again for your response and the great library!
